### DISCUSSION TITLE
Dev branch is often not-buildable

### DISCUSSION THREAD

**Discussion by emandtf** [Koenkk/zigbee2mqtt#20347] [General]
I know that DEV branch is not-stable, but **why developers do commits that prevent compiling?**
I just pulled from DEV (to check for new things and to test something), run "_npm ci_" and then "_npm start_" but I get this error:

> Error: Cannot find module './dist/util/settings'
>  Require stack:
>  /opt/zigbee2mqtt/index.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1039:15)
    at Function.Module._load (node:internal/modules/cjs/loader:885:27)
    at Module.require (node:internal/modules/cjs/loader:1105:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at start (/opt/zigbee2mqtt/index.js:90:22)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)

It's a bit frustrating because it's not possibile to test for newer things most of the times I pull from DEV....
Now I have to search for a commit where I can run 1.34-dev to be allowed to run my Conbee 3 and with the fix for "_duplicated database.db entitites_"....

Neither returning to prev commit ("2c5d3e14f779ff138949f11f6a443fd8821e0120" on DEV) seems working now.....VERY VERY FRUSTRATING

**Comment by emandtf** [comment]
Removing "_tsconfig.tsbuildinfo_" and do "npm ci && npm start" doesn't works.
The solution was to manually do "_npm run build_" after the "_tsconfig.tsbuildinfo_".

:-/

**Reply by Koenkk** [reply]
> The solution was to manually do "npm run build" after the "tsconfig.tsbuildinfo".

Do you mean after removing the `tsconfig.tsbuildinfo`?

**Reply by emandtf** [reply]
🤭 Yes, I forgot a piece.

Maybe something changed in that file when I checkout to a newer commit or run "npm ci" on newer commits so even returning back to prev working commit didn't solved anything.

**Reply by Koenkk** [reply]
@FabianMangold I still see some occasional issues with the incremental build, is this something you could check?

**Reply by FabianMangold** [reply]
Hi @Koenkk,
I sure will, as it bit me no later than yesterday !

@emandtf: What helped is the following:
- `npm run clean` -> cleans everything
- `npm ci `-> clean install -> pulls new dependencies in case required
- `npm run build `-> recompile the project (in this case a full build)

Could you please try above ?
For information, removing the buildinfo is equivalent to running `npm run clean` as a matter of fact.
This file is required for the transpiler to store the status of an incremental builds, so removing it manually will force a full build. Still removing it manually is unlikely to be the right approach.
In my case, pulling in the latest dependencies did the trick. As a matter of fact, i was working on an outdated repository, and simply pulling in the latest changes from upstream (`git fetch upstream &&  git pull upstream master --rebase`) changed the packages.json.
When attempting to build (full build or incremental) on such a codebase, this process may fail for various reasons. In my specific case, a dependency to another project (zigbee-herdsman) couldn't be resolved (following a method signature change) and caused the build failure.
Cleaning the repository (run clean), ensures that the next build is going to be a "full build". "npm ci" in turn ensures that all dependencies are up to date.
Subsequent build should then work fine.

Please let me know if that helps.
If so, we may have to review the project dependencies. I'll look into that.

Happy holidays to all of you :)
